subsys/fs/nvs/nvs.c

```diff
@@ -396,28 +394,40 @@ static int _nvs_sector_close(struct nvs_fs *fs)
 static int _nvs_gc(struct nvs_fs *fs)
 {
 	int rc;
-	struct nvs_ate gc_ate, wlk_ate;
-	u32_t gc_addr, wlk_addr, wlk_prev_addr, data_addr, sec_addr;
+	struct nvs_ate close_ate, gc_ate, wlk_ate;
+	u32_t sec_addr, gc_addr, gc_prev_addr, wlk_addr, wlk_prev_addr,
+	      data_addr, stop_addr;
 	size_t ate_size;
 
 	ate_size = _nvs_al_size(fs, sizeof(struct nvs_ate));
 
-	gc_addr = (fs->ate_wra & ADDR_SECT_MASK) + fs->sector_size - ate_size;
-	_nvs_sector_advance(fs, &gc_addr);
-	sec_addr = gc_addr;
+	sec_addr = (fs->ate_wra & ADDR_SECT_MASK);
+	_nvs_sector_advance(fs, &sec_addr);
+	gc_addr = sec_addr + fs->sector_size - ate_size;
 
-	while (1) {
-		/* if the sector is empty don't do gc */
-		rc = _nvs_flash_cmp_const(fs, gc_addr, 0xff, sizeof(struct nvs_ate));
-		if (!rc) {
-			break;
-		}
-		/* if sector end is reached stop gc */
-		rc = _nvs_flash_cmp_const(fs, gc_addr, 0x00, sizeof(struct nvs_ate));
-		if (!rc) {
-			break;
+	/* if the sector is not closed don't do gc */
+	rc = _nvs_flash_cmp_const(fs, gc_addr, 0xff, sizeof(struct nvs_ate));
+	if (!rc) {
+		rc = _nvs_flash_erase_sector(fs, sec_addr);
```

Comment:
I wonder in which case we need to garbage collect a sector that is not closed, but also not empty (as we need to erase it). This is nothing new in this patch, but just more visible as it's now out of the loop.

Reply:
In addition failing to read the closing ate of a sector (rc < 0) will erase it. I thing an error should be returned instead.

Reply:
Hi @aurel32, thank you for reviewing the changes. Normally there should be no case where we need to gc a sector that is not closed. This could only appear if something else has written to the flash location reserved for nvs. To get the system back on track it just erases the sector so that it can be used for writing in future, if it would not be empty there is a risk of writing wrong data. It could return an error (better would be a warning that some error was corrected), but at the moment I don't see a possibility to generate such a warning.

Reply:
Thanks for the details. I think that at least an error should be returned if reading the ate failed instead of erasing the flash in that case. Otherwise indeed it's probably better to just erase the sector.